src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -75,23 +77,40 @@ protected internal override async ValueTask<TRequestedArray> Read<TRequestedArra
         /// <summary>
         /// Reads an array of element type <typeparamref name="TRequestedElement"/> from the given buffer <paramref name="buf"/>.
         /// </summary>
-        protected async ValueTask<Array> ReadArray<TRequestedElement>(NpgsqlReadBuffer buf, bool async, int expectedDimensions = 0)
+        protected async ValueTask<Array> ReadArray<TRequestedElement>(NpgsqlReadBuffer buf, bool async, int expectedDimensions = 0, bool readAsObject = false)
         {
             await buf.Ensure(12, async);
             var dimensions = buf.ReadInt32();
             var containsNulls = buf.ReadInt32() == 1;
             buf.ReadUInt32(); // Element OID. Ignored.
 
-            if (ElementTypeInfo<TRequestedElement>.IsNonNullable && containsNulls)
-                throw new InvalidOperationException(ReadNonNullableCollectionWithNullsExceptionMessage);
+            var returnType = readAsObject && typeof(TRequestedElement).IsValueType
+                ? ValueTypeArrayNullability switch
+                {
+                    ValueTypeArrayNullability.Never => containsNulls
+                        ? throw new InvalidOperationException(ReadNonNullableCollectionWithNullsExceptionMessage)
+                        : typeof(TRequestedElement),
+                    ValueTypeArrayNullability.Always => typeof(Nullable<>).MakeGenericType(typeof(TRequestedElement)),
```

Comment:
We should consider calling MakeGenericType once at static initialization (or similar), and caching that.

Reply:
I was lazy here, leaving performance related things for later.
I did the caching now.
I think there are some more possible optimizations (e.g. creating one-dimensional arrays of nullable value types by calling the constructor via a compiled expression even when they are requested as object) but TBH in that case I'm not sure whether it's worthwhile).